### DISCUSSION TITLE
Einsum notation

### DISCUSSION THREAD

**Discussion by saulshanabrook** [Quansight-Labs/metadsl#128] [Ideas]
As we are thinking about translating NumPy-ish API to MoA forms to reduce, one of the most general that comes up is the einsum notation and function (https://docs.scipy.org/doc/numpy/reference/generated/numpy.einsum.html#numpy.einsum) (Note that tensorflow's einsum supports a subset of this form https://www.tensorflow.org/api_docs/python/tf/einsum) AFAIK [It is the superset of the `np.inner` and `np.tensordot`](http://numpy-discussion.10968.n7.nabble.com/einsum-slow-vs-tensor-dot-tp17418p17419.html), so if we can reduce any form of einsum, then we can reduce those as well.  Also, we would then be able to used `opt_einsum` to optimize the einsum computation first, and use uarray/moa as the backend (https://optimized-einsum.readthedocs.io/en/latest/backends.html) All we need to do is be able to get the shapes of arrays and define einsum to plug it into that library.

**Comment by saulshanabrook** [comment]
Reference for how einsum covers many other operations: https://rockt.github.io/2018/04/30/einsum

**Comment by womenflyplanes** [comment]
I'll look at einsum after we get the proposal out.

**Comment by saulshanabrook** [comment]
PR adding optimization to einsum: https://github.com/numpy/numpy/pull/5488

**Comment by saulshanabrook** [comment]
Original thread for implementation: http://numpy-discussion.10968.n7.nabble.com/einsum-td11810.html

**Comment by saulshanabrook** [comment]
I think I am getting an intuition for what this notation is and why it comes up near the term "tensor contraction".

If you are following this, I would first recommend  reading [this blog post](https://obilaniu6266h16.wordpress.com/2016/02/04/einstein-summation-in-numpy/) about einsum's notation which gives you some pretty pictures to explain it.

---

Really what we are doing is taking some incoming arrays and a result array and labeling the axis/dimensions of them. Each dimension gets some label, in NumPy these are single letters, like `i`, `j`, `k`. 

Let's think about the simplest case, where you just have one input and one output and the dimensionality of them are the same. Then we are just re-ordering the dimensions. 

```
res = einsum('ijk->kij', a)
# this is equivalent to
res[i, j, k] == a[k, i j]
```

What is this operator? In MoA it is a generalized Transpose:

<img width="669" alt="screen shot 2018-10-31 at 1 06 01 pm" src="https://user-images.githubusercontent.com/1186124/47805478-c7ba7880-dd0d-11e8-842f-0c564889e56b.png">

So an interesting implementation would be to try at first implementing the generalized Transpose with our current Sequence approach and see how that feels.

**Comment by womenflyplanes** [comment]
Agreed.

On Thu, Nov 1, 2018 at 12:05 PM Saul Shanabrook <notifications@github.com>
wrote:

> I think I am getting an intuition for what this notation is and why it
> comes up near the term "tensor contraction".
>
> If you are following this, I would first recommend reading this blog post
> <https://obilaniu6266h16.wordpress.com/2016/02/04/einstein-summation-in-numpy/>
> about einsum's notation which gives you some pretty pictures to explain it.
> ------------------------------
>
> Really what we are doing is taking some incoming arrays and a result array
> and labeling the axis/dimensions of them. Each dimension gets some label,
> in NumPy these are single letters, like i, j, k.
>
> Let's think about the simplest case, where you just have one input and one
> output and the dimensionality of them are the same. Then we are just
> re-ordering the dimensions.
>
> res = einsum('ijk->kij', a)
> # this is equivalent to
> res[i, j, k] == a[k, i j]
>
> What is this operator? In MoA it is a generalized Transpose:
>
> [image: screen shot 2018-10-31 at 1 06 01 pm]
> <https://user-images.githubusercontent.com/1186124/47805478-c7ba7880-dd0d-11e8-842f-0c564889e56b.png>
>
> So an interesting implementation would be to try at first implementing the
> generalized Transpose with our current Sequence approach and see how that
> feels.
>
> —
> You are receiving this because you commented.
> Reply to this email directly, view it on GitHub
> <https://github.com/Quansight-Labs/uarray/issues/17#issuecomment-435085701>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AGsmlO8ng0xoScsfo0y7IAYcqJUcMjo7ks5uqxuEgaJpZM4XxKD6>
> .
>
-- 
"Great spirits have always encountered violent opposition from mediocre
minds" - Albert Einstein

**Comment by womenflyplanes** [comment]
Everything else; traces, selections of dimensions, etc., can be formulated
with not only transpose, but Omega, rotate, and selection for the trace.
The entirety can be simplified to show this.
Lenore

On Thu, Nov 1, 2018 at 12:13 PM Lenore Mullin <lenore.mullin@gmail.com>
wrote:

> Agreed.
>
> On Thu, Nov 1, 2018 at 12:05 PM Saul Shanabrook <notifications@github.com>
> wrote:
>
>> I think I am getting an intuition for what this notation is and why it
>> comes up near the term "tensor contraction".
>>
>> If you are following this, I would first recommend reading this blog post
>> <https://obilaniu6266h16.wordpress.com/2016/02/04/einstein-summation-in-numpy/>
>> about einsum's notation which gives you some pretty pictures to explain it.
>> ------------------------------
>>
>> Really what we are doing is taking some incoming arrays and a result
>> array and labeling the axis/dimensions of them. Each dimension gets some
>> label, in NumPy these are single letters, like i, j, k.
>>
>> Let's think about the simplest case, where you just have one input and
>> one output and the dimensionality of them are the same. Then we are just
>> re-ordering the dimensions.
>>
>> res = einsum('ijk->kij', a)
>> # this is equivalent to
>> res[i, j, k] == a[k, i j]
>>
>> What is this operator? In MoA it is a generalized Transpose:
>>
>> [image: screen shot 2018-10-31 at 1 06 01 pm]
>> <https://user-images.githubusercontent.com/1186124/47805478-c7ba7880-dd0d-11e8-842f-0c564889e56b.png>
>>
>> So an interesting implementation would be to try at first implementing
>> the generalized Transpose with our current Sequence approach and see how
>> that feels.
>>
>> —
>> You are receiving this because you commented.
>> Reply to this email directly, view it on GitHub
>> <https://github.com/Quansight-Labs/uarray/issues/17#issuecomment-435085701>,
>> or mute the thread
>> <https://github.com/notifications/unsubscribe-auth/AGsmlO8ng0xoScsfo0y7IAYcqJUcMjo7ks5uqxuEgaJpZM4XxKD6>
>> .
>>
> --
> "Great spirits have always encountered violent opposition from mediocre
> minds" - Albert Einstein
>


-- 
"Great spirits have always encountered violent opposition from mediocre
minds" - Albert Einstein

**Comment by saulshanabrook** [comment]
There is some code in SymPy currently which converts between index expressions and tensor expressions: https://github.com/sympy/sympy/blob/c527249227b83b9f478e05629b9ffbb11f2e55b4/sympy/codegen/array_utils.py#L807-L862

This could be helpful to turn some reduced ONF into higher level operations, like an inner product, that could be compiled to optimized versions of those functions.   https://github.com/Quansight-Labs/uarray/issues/66

**Comment by saulshanabrook** [comment]
@costrouc this post might be relevant to you!
